File: framework/data-jpa/src/main/java/org/springframework/data/jpa/repository/query/MybatisParameterBinder.java
package org.springframework.data.jpa.repository.query;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;
import javax.persistence.Query;
import org.apache.ibatis.binding.MapperMethod.ParamMap;
import org.apache.ibatis.executor.ErrorContext;
import org.apache.ibatis.mapping.BoundSql;
import org.apache.ibatis.mapping.MappedStatement;
import org.apache.ibatis.mapping.ParameterMapping;
import org.apache.ibatis.mapping.ParameterMode;
import org.apache.ibatis.reflection.MetaObject;
import org.apache.ibatis.reflection.ParamNameResolver;
import org.apache.ibatis.session.Configuration;
import org.apache.ibatis.type.TypeException;
import org.apache.ibatis.type.TypeHandlerRegistry;
import org.springframework.data.jpa.repository.query.JpaParameters.JpaParameter;
import org.springframework.data.jpa.repository.query.QueryParameterSetter.BindableQuery;
import org.springframework.data.jpa.repository.query.QueryParameterSetter.ErrorHandling;
import org.springframework.data.jpa.repository.query.QueryParameterSetter.QueryMetadata;
import top.bettercode.simpleframework.data.jpa.query.mybatis.MybatisParam;
import top.bettercode.simpleframework.data.jpa.support.Size;

/**
 * @author Peter Wu
 */
public class MybatisParameterBinder extends ParameterBinder {

  private static final String GENERIC_NAME_PREFIX = "param";

  private final JpaParameters parameters;
  private final TypeHandlerRegistry typeHandlerRegistry;
  private final MappedStatement mappedStatement;
  private final Configuration configuration;

  public MybatisParameterBinder(JpaParameters parameters,
      MappedStatement mappedStatement) {
    super(parameters, Collections.emptyList());
    this.parameters = parameters;
    this.mappedStatement = mappedStatement;
    this.configuration = mappedStatement.getConfiguration();
    this.typeHandlerRegistry = mappedStatement.getConfiguration().getTypeHandlerRegistry();
  }

  @Deprecated
  @Override
  public <T extends Query> T bind(T jpaQuery, QueryMetadata metadata,
      JpaParametersParameterAccessor accessor) {
    return jpaQuery;
  }

  @Deprecated
  @Override
  public void bind(BindableQuery query, JpaParametersParameterAccessor accessor,
      ErrorHandling errorHandling) {
  }


  @Deprecated
  @Override
  Query bindAndPrepare(Query query, QueryMetadata metadata,
      JpaParametersParameterAccessor accessor) {
    return query;
  }

  void bind(BindableQuery query, MybatisParam mybatisParam) {
    ErrorContext.instance().activity("setting parameters")
        .object(mappedStatement.getParameterMap().getId());
    ErrorHandling errorHandling = ErrorHandling.STRICT;

    BoundSql boundSql = mybatisParam.getBoundSql();
    Object parameterObject = mybatisParam.getParameterObject();
    List<ParameterMapping> parameterMappings = boundSql.getParameterMappings();
    if (parameterMappings != null) {
      for (int i = 0; i < parameterMappings.size(); i++) {
        ParameterMapping parameterMapping = parameterMappings.get(i);
        if (parameterMapping.getMode() != ParameterMode.OUT) {
          Object value;
          String propertyName = parameterMapping.getProperty();
          if (boundSql.hasAdditionalParameter(propertyName)) {
            value = boundSql.getAdditionalParameter(propertyName);
          } else if (parameterObject == null) {
            value = null;
          } else if (typeHandlerRegistry.hasTypeHandler(parameterObject.getClass())) {
            value = parameterObject;
          } else {
            MetaObject metaObject = configuration.newMetaObject(parameterObject);
            value = metaObject.getValue(propertyName);
          }
          try {
            int position = i + 1;
            errorHandling.execute(() -> query.setParameter(position, value));
          } catch (Exception e) {
            throw new TypeException(
                "Could not set parameters for mapping: " + parameterMapping + ". Cause: " + e, e);
          }
        }
      }
    }

  }


  Query bindAndPrepare(Query query,QueryParameterSetter.QueryMetadata metadata, JpaParametersParameterAccessor accessor,
      MybatisParam mybatisParam) {

    bind(metadata.withQuery(query), mybatisParam);

    Size size = mybatisParam.getSize();

    if (size != null) {
      query.setFirstResult(0);
      query.setMaxResults(size.getSize());
    }

    if (!parameters.hasPageableParameter() || accessor.getPageable().isUnpaged()) {
      return query;
    }

    query.setFirstResult((int) accessor.getPageable().getOffset());
    query.setMaxResults(accessor.getPageable().getPageSize());

    return query;
  }


  public MybatisParam bindParameterObject(JpaParametersParameterAccessor accessor) {
    Size size = null;
    Object[] values = accessor.getValues();
    int bindableSize = 0;
    final Map<String, Object> paramMap = new ParamMap<>();
    Set<String> names = this.parameters.stream().map(p -> p.getName().orElse(null))
        .filter(Objects::nonNull).collect(Collectors.toSet());
    Object params = null;
    for (JpaParameter parameter : this.parameters.getBindableParameters()) {
      Class<?> parameterType = parameter.getType();
      int parameterIndex = parameter.getIndex();
      Object value = values[parameterIndex];
      if (Size.class.isAssignableFrom(parameterType)) {
        size = (Size) value;
      } else {
        Optional<String> name = parameter.getName();
        name.ifPresent(s -> paramMap.put(s, value));

        String otherName = GENERIC_NAME_PREFIX + (bindableSize + 1);
        if (!names.contains(otherName)) {
          paramMap.put(otherName, value);
        }

        if (bindableSize == 0) {
          params = ParamNameResolver.wrapToMapIfCollection(value, name.orElse(null));
        }

        bindableSize++;
      }
    }
    params = bindableSize > 1 ? paramMap : params;
    Object parameterObject = params;
    BoundSql boundSql = mappedStatement.getBoundSql(parameterObject);
    return new MybatisParam(boundSql, parameterObject, size);
  }

}

File: data-jpa/src/main/kotlin/org/springframework/data/jpa/repository/query/MybatisParameterBinder.kt
package org.springframework.data.jpa.repository.query

import org.apache.ibatis.binding.MapperMethod.ParamMap
import org.apache.ibatis.executor.ErrorContext
import org.apache.ibatis.mapping.MappedStatement
import org.apache.ibatis.mapping.ParameterMode
import org.apache.ibatis.reflection.ParamNameResolver
import org.apache.ibatis.session.Configuration
import org.apache.ibatis.type.TypeException
import org.apache.ibatis.type.TypeHandlerRegistry
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.query.JpaParameters.JpaParameter
import org.springframework.data.jpa.repository.query.QueryParameterSetter.BindableQuery
import org.springframework.data.jpa.repository.query.QueryParameterSetter.ErrorHandling
import top.bettercode.summer.data.jpa.query.mybatis.MybatisParam
import top.bettercode.summer.data.jpa.support.Size
import java.util.*
import javax.persistence.Query

/**
 * @author Peter Wu
 */
internal class MybatisParameterBinder(
    private val parameters: JpaParameters,
    private val paramed: Boolean,
    private val mappedStatement: MappedStatement
) : ParameterBinder(parameters, emptyList()) {
    private val typeHandlerRegistry: TypeHandlerRegistry =
        mappedStatement.configuration.typeHandlerRegistry
    private val configuration: Configuration = mappedStatement.configuration

    override fun <T : Query> bind(
        jpaQuery: T,
        metadata: QueryParameterSetter.QueryMetadata,
        accessor: JpaParametersParameterAccessor
    ): T {
        return jpaQuery
    }

    @Deprecated("")
    override fun bind(
        query: BindableQuery, accessor: JpaParametersParameterAccessor,
        errorHandling: ErrorHandling
    ) {
    }

    public override fun bindAndPrepare(
        query: Query, metadata: QueryParameterSetter.QueryMetadata,
        accessor: JpaParametersParameterAccessor
    ): Query {
        return query
    }

    fun bind(query: BindableQuery, mybatisParam: MybatisParam) {
        ErrorContext.instance().activity("setting parameters")
            .`object`(mappedStatement.parameterMap.id)
        val errorHandling = ErrorHandling.STRICT
        val boundSql = mybatisParam.boundSql
        val parameterObject = mybatisParam.parameterObject
        val parameterMappings = boundSql.parameterMappings
        if (parameterMappings != null) {
            for (i in parameterMappings.indices) {
                val parameterMapping = parameterMappings[i]
                if (parameterMapping.mode != ParameterMode.OUT) {
                    var value: Any?
                    val propertyName = parameterMapping.property
                    value = if (boundSql.hasAdditionalParameter(propertyName)) {
                        boundSql.getAdditionalParameter(propertyName)
                    } else if (parameterObject == null) {
                        null
                    } else if (typeHandlerRegistry.hasTypeHandler(parameterObject.javaClass)) {
                        parameterObject
                    } else {
                        val metaObject = configuration.newMetaObject(parameterObject)
                        metaObject.getValue(propertyName)
                    }
                    try {
                        val position = i + 1
                        errorHandling.execute {
                            @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
                            query.setParameter(position, value)
                        }
                    } catch (e: Exception) {
                        throw TypeException(
                            "Could not set parameters for mapping: $parameterMapping. Cause: $e", e
                        )
                    }
                }
            }
        }
    }

    fun bindAndPrepare(
        query: Query,
        metadata: QueryParameterSetter.QueryMetadata,
        pageable: Pageable,
        mybatisParam: MybatisParam
    ): Query {
        bind(metadata.withQuery(query), mybatisParam)
        val size = mybatisParam.size
        if (size != null && !size.isUnlimited()) {
            query.setFirstResult(0)
            query.setMaxResults(size.size)
        }
        if (!parameters.hasPageableParameter() || pageable.isUnpaged) {
            return query
        }
        query.setFirstResult(pageable.offset.toInt())
        query.setMaxResults(pageable.pageSize)
        return query
    }

    fun bindParameterObject(accessor: JpaParametersParameterAccessor): MybatisParam {
        var size: Size? = null
        val values = accessor.values
        var bindableSize = 0
        val paramMap: MutableMap<String?, Any?> = ParamMap()
        val names = this.parameters.map { p: JpaParameter -> p.name.orElse(null) }
            .filter { obj: String? -> Objects.nonNull(obj) }.toSet()
        var params: Any? = null
        for (parameter in this.parameters.bindableParameters) {
            val parameterType = parameter.type
            val parameterIndex = parameter.index
            val value = values[parameterIndex]
            if (Size::class.java.isAssignableFrom(parameterType)) {
                size = value as Size?
            } else {
                val name = parameter.name
                name.ifPresent { s: String? -> paramMap[s] = value }
                val otherName = GENERIC_NAME_PREFIX + (bindableSize + 1)
                if (!names.contains(otherName)) {
                    paramMap[otherName] = value
                }
                if (bindableSize == 0) {
                    params = ParamNameResolver.wrapToMapIfCollection(value, name.orElse(null))
                }
                bindableSize++
            }
        }
        params = if (paramed || bindableSize > 1) paramMap else params
        val parameterObject = params
        val boundSql = mappedStatement.getBoundSql(parameterObject)
        return MybatisParam(boundSql, parameterObject, size)
    }

    companion object {
        private const val GENERIC_NAME_PREFIX = "param"
    }
}
